File: 301.Remove_Invalid_Parentheses/Remove_Invalid_Parentheses.cpp
#include<iostream>
#include<vector>
#include<string>
#include<unordered_set>
using namespace std;

class Solution {
public:
    vector<string> removeInvalidParentheses(string s) {
        unordered_set<string> queue;
        std::vector<string> res;
        queue.insert(s);
        while(true){
        	for(auto candidate: queue){
        		if(isValid(candidate)){
        		    cout<<candidate<<"in"<<endl;
        			res.push_back(candidate);
        		}
        	}
        	if(res.size() != 0){
        		return res;
        	}
        	
            unordered_set<string> t;
        	for(auto candidate: queue){
        		for(int i=0; i<candidate.size();i++){
        			string tmp = candidate.substr(0,i) + candidate.substr(i+1);
        			t.insert(tmp);
        		}
        	}
        	queue = t;
        }
    }

    bool isValid(string s){
    	int Sum = 0;
    	if(s == "") return true;
    	for(int i=0; i<s.size(); i++){
    		if(s[i] == '(')	
    			Sum++;
    		if(s[i] == ')') 
    			Sum--;
    		if(Sum < 0) 
    			return false;
    	}
    	return Sum == 0;
    }
};
int main(){
	Solution s;
	vector<string> res = s.removeInvalidParentheses(")(");
	cout<< res.size() << endl;
	for(auto c: res){
		std::cout<<c<<std::endl;
	}
}

File: 301.Remove_Invalid_Parentheses/Remove_Invalid_Parentheses.py
class Solution(object):
    def removeInvalidParentheses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        def isValid(s):
        	dic = {'(':1, ')':-1}
        	Sum = 0
        	for ch in s:
        		Sum += dic.get(ch, 0)
        		if Sum < 0:
        			return False
        	return Sum == 0
        queue = {s}
        res = []
        while True:
	        for candidate in queue:
	        	if isValid(candidate):
	        		res.append(candidate)
	        if res:
	        	return res
	        queue = {s[:i]+s[i+1:] for s in queue for i in range(len(s))}


       	# def isValid(s):
        #     a = 0
        #     for ch in s:
        #         a += {'(':1, ')':-1}.get(ch, 0)
        #         if a < 0:
        #             return False
        #     return a == 0
        # queue = {s}
        # while True:
        #     res = filter(isValid, queue)
        #     if res:
        #         return res
        #     queue = {s[:i]+s[i+1:] for s in queue for i in range(len(s))}
        # return res


